
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Introduction to scvi-tools &#8212; Extended single-cell analysis tutorial</title>
    
  <link href="../_static/css/theme.css" rel="stylesheet">
  <link href="../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/book.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/togglebutton.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/book.js"></script>
    <script async="async" kind="hypothesis" src="https://hypothes.is/embed.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../_static/logo.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">Extended single-cell analysis tutorial</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Introduction
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="prior_art.html">
   1. Prior art
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="experimental_data_collection.html">
   2. Experimental data collection
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="raw_data_processing.html">
   3. Raw data processing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="analysis_tools.html">
   4. Computational single-cell analysis frameworks and tools
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="data_infrastructure.html">
   5. Data infrastructure
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="interoperability.html">
   6. Interoperability
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Demo
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../scanpy_demo.html">
   7. Scanpy demo to test the environment
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../_sources/introduction/api_overview.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/theislab/extended-single-cell-tutorial"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/theislab/extended-single-cell-tutorial/issues/new?title=Issue%20on%20page%20%2Fintroduction/api_overview.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/theislab/extended-single-cell-tutorial/edit/master/introduction/api_overview.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Launch interactive content"><i class="fas fa-rocket"></i></button>
    <div class="dropdown-buttons">
        
        <a class="binder-button" href="https://mybinder.org/v2/gh/theislab/extended-single-cell-tutorial/master?urlpath=tree/introduction/api_overview.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Launch Binder" data-toggle="tooltip"
                data-placement="left"><img class="binder-button-logo"
                    src="../_static/images/logo_binder.svg"
                    alt="Interact on binder">Binder</button></a>
        
        
        
        
    </div>
</div>

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#loading-and-preparing-data">
   Loading and preparing data
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#creating-and-training-a-model">
   Creating and training a model
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#saving-and-loading">
     Saving and loading
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#obtaining-model-outputs">
   Obtaining model outputs
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#interoperability-with-scanpy">
   Interoperability with Scanpy
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#visualization-without-batch-correction">
     Visualization without batch correction
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#visualization-with-batch-correction-scvi">
     Visualization with batch correction (scVI)
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#clustering-on-the-scvi-latent-space">
     Clustering on the scVI latent space
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#differential-expression">
   Differential expression
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#logging-information">
   Logging information
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="introduction-to-scvi-tools">
<h1>Introduction to scvi-tools<a class="headerlink" href="#introduction-to-scvi-tools" title="Permalink to this headline">¶</a></h1>
<p>In this introductory tutorial, we go through the different steps of an scvi-tools workflow.</p>
<p>While we focus on scVI in this tutorial, the API is consistent across all models.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="o">!</span>pip install --quiet scvi-colab
<span class="kn">from</span> <span class="nn">scvi_colab</span> <span class="kn">import</span> <span class="n">install</span>
<span class="n">install</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">scvi</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>

<span class="n">sc</span><span class="o">.</span><span class="n">set_figure_params</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">))</span>

<span class="c1"># for white background of figures (only for docs rendering)</span>
<span class="o">%</span><span class="k">config</span> InlineBackend.print_figure_kwargs={&#39;facecolor&#39; : &quot;w&quot;}
<span class="o">%</span><span class="k">config</span> InlineBackend.figure_format=&#39;retina&#39;
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Global seed set to 0
</pre></div>
</div>
</div>
</div>
<div class="section" id="loading-and-preparing-data">
<h2>Loading and preparing data<a class="headerlink" href="#loading-and-preparing-data" title="Permalink to this headline">¶</a></h2>
<p>Let us first load a subsampled version of the heart cell atlas dataset described in Litviňuková et al. (2020). scvi-tools has many “built-in” datasets as well as support for loading arbitrary <code class="docutils literal notranslate"><span class="pre">.csv</span></code>, <code class="docutils literal notranslate"><span class="pre">.loom</span></code>, and <code class="docutils literal notranslate"><span class="pre">.h5ad</span></code> (AnnData) files. Please see our tutorial on data loading for more examples.</p>
<ul class="simple">
<li><p>Litviňuková, M., Talavera-López, C., Maatz, H., Reichart, D., Worth, C. L., Lindberg, E. L., … &amp; Teichmann, S. A. (2020). Cells of the adult human heart. Nature, 588(7838), 466-472.</p></li>
</ul>
<div class="alert alert-info">
Important
<p>All scvi-tools models require AnnData objects as input.</p>
</div><div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span> <span class="o">=</span> <span class="n">scvi</span><span class="o">.</span><span class="n">data</span><span class="o">.</span><span class="n">heart_cell_atlas_subsampled</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Blue">INFO    </span> File data/hca_subsampled_20k.h5ad already downloaded                                
</pre></div>
</div>
</div>
</div>
<p>Now we preprocess the data to remove, for example, genes that are very lowly expressed and other outliers. For these tasks we prefer the <a class="reference external" href="https://scanpy.readthedocs.io/en/stable/api/index.html#module-scanpy.pp">Scanpy preprocessing module</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_counts</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>In scRNA-seq analysis, it’s popular to normalize the data. These values are not used by scvi-tools, but given their popularity in other tasks as well as for visualization, we store them in the anndata object separately (via the <code class="docutils literal notranslate"><span class="pre">.raw</span></code> attribute).</p>
<div class="alert alert-info">
<p>Important</p>
<p>Unless otherwise specified, scvi-tools models require the raw counts (not log library size normalized).</p>
</div><div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;counts&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span> <span class="c1"># preserve counts</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">normalize_total</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">target_sum</span><span class="o">=</span><span class="mf">1e4</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">log1p</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">adata</span><span class="o">.</span><span class="n">raw</span> <span class="o">=</span> <span class="n">adata</span> <span class="c1"># freeze the state in `.raw`</span>
</pre></div>
</div>
</div>
</div>
<p>Finally, we perform feature selection, to reduce the number of features (genes in this case) used as input to the scvi-tools model. For best practices of how/when to perform feature selection, please refer to the model-specific tutorial. For scVI, we recommend anywhere from 1,000 to 10,000 HVGs, but it will be context-dependent.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">highly_variable_genes</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span> 
    <span class="n">n_top_genes</span><span class="o">=</span><span class="mi">1200</span><span class="p">,</span> 
    <span class="n">subset</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> 
    <span class="n">layer</span><span class="o">=</span><span class="s2">&quot;counts&quot;</span><span class="p">,</span> 
    <span class="n">flavor</span><span class="o">=</span><span class="s2">&quot;seurat_v3&quot;</span><span class="p">,</span>
    <span class="n">batch_key</span><span class="o">=</span><span class="s2">&quot;cell_source&quot;</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>Now it’s time to run <code class="docutils literal notranslate"><span class="pre">setup_anndata()</span></code>, which alerts scvi-tools to the locations of various matrices inside the anndata. It’s important to run this function with the correct arguments so scvi-tools is notified that your dataset has batches, annotations, etc. For example, if batches are registered with scvi-tools, the subsequent model will correct for batch effects. See the full documentation for details.</p>
<p>In this dataset, there is a “cell_source” categorical covariate, and within each “cell_source”, multiple “donors”, “gender” and “age_group”. There are also two continuous covariates we’d like to correct for: “percent_mito” and “percent_ribo”. These covariates can be registered using the <code class="docutils literal notranslate"><span class="pre">categorical_covariate_keys</span></code> argument. If you only have one categorical covariate, you can also use the <code class="docutils literal notranslate"><span class="pre">batch_key</span></code> argument instead.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">scvi</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">SCVI</span><span class="o">.</span><span class="n">setup_anndata</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span>
    <span class="n">layer</span><span class="o">=</span><span class="s2">&quot;counts&quot;</span><span class="p">,</span>
    <span class="n">categorical_covariate_keys</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;cell_source&quot;</span><span class="p">,</span> <span class="s2">&quot;donor&quot;</span><span class="p">],</span>
    <span class="n">continuous_covariate_keys</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;percent_mito&quot;</span><span class="p">,</span> <span class="s2">&quot;percent_ribo&quot;</span><span class="p">]</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="alert alert-warning">
<p>Warning</p>
<p>If the adata is modified after running <code class="docutils literal notranslate"><span class="pre">setup_anndata</span></code>, please run <code class="docutils literal notranslate"><span class="pre">setup_anndata</span></code> again, before creating an instance of a model.</p>
</div></div>
<div class="section" id="creating-and-training-a-model">
<h2>Creating and training a model<a class="headerlink" href="#creating-and-training-a-model" title="Permalink to this headline">¶</a></h2>
<p>While we highlight the scVI model here, the API is consistent across all scvi-tools models and is inspired by that of <a class="reference external" href="https://scikit-learn.org/stable/">scikit-learn</a>. For a full list of options, see the scvi <a class="reference external" href="https://scvi-tools.org">documentation</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">model</span> <span class="o">=</span> <span class="n">scvi</span><span class="o">.</span><span class="n">model</span><span class="o">.</span><span class="n">SCVI</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<p>We can see an overview of the model by printing it.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">model</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><pre style="white-space:pre;overflow-x:auto;line-height:normal;font-family:Menlo,'DejaVu Sans Mono',consolas,'Courier New',monospace">SCVI Model with the following params: 
n_hidden: <span style="color: #008080; text-decoration-color: #008080; font-weight: bold">128</span>, n_latent: <span style="color: #008080; text-decoration-color: #008080; font-weight: bold">10</span>, n_layers: <span style="color: #008080; text-decoration-color: #008080; font-weight: bold">1</span>, dropout_rate: <span style="color: #008080; text-decoration-color: #008080; font-weight: bold">0.1</span>, dispersion: gene, 
gene_likelihood: zinb, latent_distribution: normal
Training status: Not Trained
</pre>
</div><div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
</div>
<div class="alert alert-info">
Important
<p>All scvi-tools models run faster when using a GPU. By default, scvi-tools will use a GPU if one is found to be available. Please see the installation page for more information about installing scvi-tools when a GPU is available.</p>
</div><div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">model</span><span class="o">.</span><span class="n">train</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>GPU available: True, used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>Epoch 400/400: 100%|██████████| 400/400 [05:43&lt;00:00,  1.16it/s, loss=284, v_num=1]
</pre></div>
</div>
</div>
</div>
<div class="section" id="saving-and-loading">
<h3>Saving and loading<a class="headerlink" href="#saving-and-loading" title="Permalink to this headline">¶</a></h3>
<p>Saving consists of saving the model neural network weights, as well as parameters used to initialize the model.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># model.save(&quot;my_model/&quot;)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># model = scvi.model.SCVI.load(&quot;my_model/&quot;, adata=adata, use_gpu=True)</span>
</pre></div>
</div>
</div>
</div>
</div>
</div>
<div class="section" id="obtaining-model-outputs">
<h2>Obtaining model outputs<a class="headerlink" href="#obtaining-model-outputs" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">latent</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">get_latent_representation</span><span class="p">()</span>
</pre></div>
</div>
</div>
</div>
<p>It’s often useful to store the outputs of scvi-tools back into the original anndata, as it permits interoperability with Scanpy.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;X_scVI&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">latent</span>
</pre></div>
</div>
</div>
</div>
<p>The <code class="docutils literal notranslate"><span class="pre">model.get...()</span></code> functions default to using the anndata that was used to initialize the model. It’s possible to also query a subset of the anndata, or even use a completely independent anndata object as long as the anndata is organized in an equivalent fashion.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_subset</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">cell_type</span> <span class="o">==</span> <span class="s2">&quot;Fibroblast&quot;</span><span class="p">]</span>
<span class="n">latent_subset</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">get_latent_representation</span><span class="p">(</span><span class="n">adata_subset</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Blue">INFO    </span> Received view of anndata, making copy.                                              
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>INFO:scvi.model.base._base_model:Received view of anndata, making copy.
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Blue">INFO    </span> Input AnnData not setup with scvi-tools. attempting to transfer AnnData setup       
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>INFO:scvi.model.base._base_model:Input AnnData not setup with scvi-tools. attempting to transfer AnnData setup
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">denoised</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">get_normalized_expression</span><span class="p">(</span><span class="n">adata_subset</span><span class="p">,</span> <span class="n">library_size</span><span class="o">=</span><span class="mf">1e4</span><span class="p">)</span>
<span class="n">denoised</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:</span><span class="mi">5</span><span class="p">,</span> <span class="p">:</span><span class="mi">5</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Blue">INFO    </span> Received view of anndata, making copy.                                              
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>INFO:scvi.model.base._base_model:Received view of anndata, making copy.
</pre></div>
</div>
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Blue">INFO    </span> Input AnnData not setup with scvi-tools. attempting to transfer AnnData setup       
</pre></div>
</div>
<div class="output stderr highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>INFO:scvi.model.base._base_model:Input AnnData not setup with scvi-tools. attempting to transfer AnnData setup
</pre></div>
</div>
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>ISG15</th>
      <th>TNFRSF18</th>
      <th>VWA1</th>
      <th>HES5</th>
      <th>SPSB1</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>GTCAAGTCATGCCACG-1-HCAHeart7702879</th>
      <td>0.343808</td>
      <td>0.118372</td>
      <td>1.945633</td>
      <td>0.062702</td>
      <td>4.456603</td>
    </tr>
    <tr>
      <th>GAGTCATTCTCCGTGT-1-HCAHeart8287128</th>
      <td>1.552080</td>
      <td>0.275485</td>
      <td>1.457585</td>
      <td>0.013442</td>
      <td>14.617260</td>
    </tr>
    <tr>
      <th>CCTCTGATCGTGACAT-1-HCAHeart7702881</th>
      <td>5.157080</td>
      <td>0.295140</td>
      <td>1.195748</td>
      <td>0.143792</td>
      <td>2.908867</td>
    </tr>
    <tr>
      <th>CGCCATTCATCATCTT-1-H0035_apex</th>
      <td>0.352172</td>
      <td>0.019281</td>
      <td>0.570386</td>
      <td>0.105633</td>
      <td>6.325405</td>
    </tr>
    <tr>
      <th>TCGTAGAGTAGGACTG-1-H0015_septum</th>
      <td>0.290155</td>
      <td>0.040910</td>
      <td>0.400771</td>
      <td>0.723409</td>
      <td>8.142258</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<p>Let’s store the normalized values back in the anndata.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">layers</span><span class="p">[</span><span class="s2">&quot;scvi_normalized&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">get_normalized_expression</span><span class="p">(</span>
    <span class="n">library_size</span><span class="o">=</span><span class="mf">10e4</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</div>
<div class="section" id="interoperability-with-scanpy">
<h2>Interoperability with Scanpy<a class="headerlink" href="#interoperability-with-scanpy" title="Permalink to this headline">¶</a></h2>
<p>Scanpy is a powerful python library for visualization and downstream analysis of scRNA-seq data. We show here how to feed the objects produced by scvi-tools into a scanpy workflow.</p>
<div class="section" id="visualization-without-batch-correction">
<h3>Visualization without batch correction<a class="headerlink" href="#visualization-without-batch-correction" title="Permalink to this headline">¶</a></h3>
<div class="alert alert-warning">
<p>Warning</p>
<p>We use UMAP to <em>qualitatively</em> assess our low-dimension embeddings of cells. We do not advise using UMAP or any similar approach quantitatively. We do recommend using the embeddings produced by scVI as a plug-in replacement of what you would get from PCA, as we show below.</p>
</div><p>First, we demonstrate the presence of nuisance variation with respect to nuclei/whole cell, age group, and donor by plotting the UMAP results of the top 30 PCA components for the raw count data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># run PCA then generate UMAP plots</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">pca</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_pcs</span><span class="o">=</span><span class="mi">30</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">20</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_dist</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span> 
    <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">],</span> 
    <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span> 
    <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;donor&quot;</span><span class="p">,</span> <span class="s2">&quot;cell_source&quot;</span><span class="p">],</span>
    <span class="n">ncols</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
    <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/api_overview_40_0.png" src="../_images/api_overview_40_0.png" />
<img alt="../_images/api_overview_40_1.png" src="../_images/api_overview_40_1.png" />
</div>
</div>
<p>We see that while the cell types are generally well separated, nuisance variation plays a large part in the variation of the data.</p>
</div>
<div class="section" id="visualization-with-batch-correction-scvi">
<h3>Visualization with batch correction (scVI)<a class="headerlink" href="#visualization-with-batch-correction-scvi" title="Permalink to this headline">¶</a></h3>
<p>Now, let us try using the scVI latent space to generate the same UMAP plots to see if scVI successfully accounts for batch effects in the data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># use scVI latent space for UMAP generation</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="s2">&quot;X_scVI&quot;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_dist</span><span class="o">=</span><span class="mf">0.3</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span> 
    <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;cell_type&quot;</span><span class="p">],</span> 
    <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span> 
    <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;donor&quot;</span><span class="p">,</span> <span class="s2">&quot;cell_source&quot;</span><span class="p">],</span>
    <span class="n">ncols</span><span class="o">=</span><span class="mi">2</span><span class="p">,</span>
    <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/api_overview_45_0.png" src="../_images/api_overview_45_0.png" />
<img alt="../_images/api_overview_45_1.png" src="../_images/api_overview_45_1.png" />
</div>
</div>
<p>We can see that scVI was able to correct for nuisance variation due to nuclei/whole cell, age group, and donor, while maintaining separation of cell types.</p>
</div>
<div class="section" id="clustering-on-the-scvi-latent-space">
<h3>Clustering on the scVI latent space<a class="headerlink" href="#clustering-on-the-scvi-latent-space" title="Permalink to this headline">¶</a></h3>
<p>The user will note that we imported curated labels from the original publication. Our interface with scanpy makes it easy to cluster the data with scanpy from scVI’s latent space and then reinject them into scVI (e.g., for differential expression).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># neighbors were already computed using scVI</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">leiden</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">key_added</span><span class="o">=</span><span class="s2">&quot;leiden_scVI&quot;</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mf">0.5</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span> 
    <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;leiden_scVI&quot;</span><span class="p">],</span> 
    <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/api_overview_50_0.png" src="../_images/api_overview_50_0.png" />
</div>
</div>
</div>
</div>
<div class="section" id="differential-expression">
<h2>Differential expression<a class="headerlink" href="#differential-expression" title="Permalink to this headline">¶</a></h2>
<p>We can also use many scvi-tools models for differential expression. For further details on the methods underlying these functions as well as additional options, please see the <a class="reference external" href="https://docs.scvi-tools.org/en/stable/api/reference/scvi.model.SCVI.differential_expression.html#scvi.model.SCVI.differential_expression">API docs</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">cell_type</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_plain highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>AACTCCCCACGAGAGT-1-HCAHeart7844001                      Myeloid
ATAACGCAGAGCTGGT-1-HCAHeart7829979    Ventricular_Cardiomyocyte
GTCAAGTCATGCCACG-1-HCAHeart7702879                   Fibroblast
GGTGATTCAAATGAGT-1-HCAHeart8102858                  Endothelial
AGAGAATTCTTAGCAG-1-HCAHeart8102863                  Endothelial
Name: cell_type, dtype: category
Categories (11, object): [&#39;Adipocytes&#39;, &#39;Atrial_Cardiomyocyte&#39;, &#39;Endothelial&#39;, &#39;Fibroblast&#39;, ..., &#39;Neuronal&#39;, &#39;Pericytes&#39;, &#39;Smooth_muscle_cells&#39;, &#39;Ventricular_Cardiomyocyte&#39;]
</pre></div>
</div>
</div>
</div>
<p>For example, a 1-vs-1 DE test is as simple as:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">de_df</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">differential_expression</span><span class="p">(</span>
    <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;cell_type&quot;</span><span class="p">,</span> 
    <span class="n">group1</span><span class="o">=</span><span class="s2">&quot;Endothelial&quot;</span><span class="p">,</span> 
    <span class="n">group2</span><span class="o">=</span><span class="s2">&quot;Fibroblast&quot;</span>
<span class="p">)</span>
<span class="n">de_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>DE...: 100%|██████████| 1/1 [00:00&lt;00:00,  3.07it/s]
</pre></div>
</div>
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>proba_de</th>
      <th>proba_not_de</th>
      <th>bayes_factor</th>
      <th>scale1</th>
      <th>scale2</th>
      <th>pseudocounts</th>
      <th>delta</th>
      <th>lfc_mean</th>
      <th>lfc_median</th>
      <th>lfc_std</th>
      <th>...</th>
      <th>raw_mean1</th>
      <th>raw_mean2</th>
      <th>non_zeros_proportion1</th>
      <th>non_zeros_proportion2</th>
      <th>raw_normalized_mean1</th>
      <th>raw_normalized_mean2</th>
      <th>is_de_fdr_0.05</th>
      <th>comparison</th>
      <th>group1</th>
      <th>group2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>SOX17</th>
      <td>0.9998</td>
      <td>0.0002</td>
      <td>8.516943</td>
      <td>0.001615</td>
      <td>0.000029</td>
      <td>0.0</td>
      <td>0.25</td>
      <td>6.222365</td>
      <td>6.216846</td>
      <td>1.967564</td>
      <td>...</td>
      <td>0.784371</td>
      <td>0.006541</td>
      <td>0.307617</td>
      <td>0.004497</td>
      <td>17.128170</td>
      <td>0.185868</td>
      <td>True</td>
      <td>Endothelial vs Fibroblast</td>
      <td>Endothelial</td>
      <td>Fibroblast</td>
    </tr>
    <tr>
      <th>SLC9A3R2</th>
      <td>0.9996</td>
      <td>0.0004</td>
      <td>7.823621</td>
      <td>0.010660</td>
      <td>0.000171</td>
      <td>0.0</td>
      <td>0.25</td>
      <td>5.977907</td>
      <td>6.049340</td>
      <td>1.672150</td>
      <td>...</td>
      <td>4.451492</td>
      <td>0.045380</td>
      <td>0.712339</td>
      <td>0.034342</td>
      <td>111.582703</td>
      <td>1.657324</td>
      <td>True</td>
      <td>Endothelial vs Fibroblast</td>
      <td>Endothelial</td>
      <td>Fibroblast</td>
    </tr>
    <tr>
      <th>ABCA10</th>
      <td>0.9990</td>
      <td>0.0010</td>
      <td>6.906745</td>
      <td>0.000081</td>
      <td>0.006355</td>
      <td>0.0</td>
      <td>0.25</td>
      <td>-8.468659</td>
      <td>-9.058912</td>
      <td>2.959383</td>
      <td>...</td>
      <td>0.014359</td>
      <td>1.282508</td>
      <td>0.007788</td>
      <td>0.499591</td>
      <td>0.501825</td>
      <td>69.613876</td>
      <td>True</td>
      <td>Endothelial vs Fibroblast</td>
      <td>Endothelial</td>
      <td>Fibroblast</td>
    </tr>
    <tr>
      <th>EGFL7</th>
      <td>0.9986</td>
      <td>0.0014</td>
      <td>6.569875</td>
      <td>0.008471</td>
      <td>0.000392</td>
      <td>0.0</td>
      <td>0.25</td>
      <td>4.751251</td>
      <td>4.730982</td>
      <td>1.546327</td>
      <td>...</td>
      <td>2.376779</td>
      <td>0.036795</td>
      <td>0.741543</td>
      <td>0.025756</td>
      <td>89.507553</td>
      <td>1.169474</td>
      <td>True</td>
      <td>Endothelial vs Fibroblast</td>
      <td>Endothelial</td>
      <td>Fibroblast</td>
    </tr>
    <tr>
      <th>VWF</th>
      <td>0.9984</td>
      <td>0.0016</td>
      <td>6.436144</td>
      <td>0.014278</td>
      <td>0.000553</td>
      <td>0.0</td>
      <td>0.25</td>
      <td>5.013347</td>
      <td>5.029471</td>
      <td>1.758744</td>
      <td>...</td>
      <td>5.072563</td>
      <td>0.054375</td>
      <td>0.808226</td>
      <td>0.032298</td>
      <td>169.693512</td>
      <td>2.207696</td>
      <td>True</td>
      <td>Endothelial vs Fibroblast</td>
      <td>Endothelial</td>
      <td>Fibroblast</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 22 columns</p>
</div></div></div>
</div>
<p>We can also do a 1-vs-all DE test, which compares each cell type with the rest of the dataset:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">de_df</span> <span class="o">=</span> <span class="n">model</span><span class="o">.</span><span class="n">differential_expression</span><span class="p">(</span>
    <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;cell_type&quot;</span><span class="p">,</span> 
<span class="p">)</span>
<span class="n">de_df</span><span class="o">.</span><span class="n">head</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>DE...: 100%|██████████| 11/11 [00:03&lt;00:00,  3.08it/s]
</pre></div>
</div>
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>proba_de</th>
      <th>proba_not_de</th>
      <th>bayes_factor</th>
      <th>scale1</th>
      <th>scale2</th>
      <th>pseudocounts</th>
      <th>delta</th>
      <th>lfc_mean</th>
      <th>lfc_median</th>
      <th>lfc_std</th>
      <th>...</th>
      <th>raw_mean1</th>
      <th>raw_mean2</th>
      <th>non_zeros_proportion1</th>
      <th>non_zeros_proportion2</th>
      <th>raw_normalized_mean1</th>
      <th>raw_normalized_mean2</th>
      <th>is_de_fdr_0.05</th>
      <th>comparison</th>
      <th>group1</th>
      <th>group2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>CIDEC</th>
      <td>0.9988</td>
      <td>0.0012</td>
      <td>6.724225</td>
      <td>0.002336</td>
      <td>0.000031</td>
      <td>0.0</td>
      <td>0.25</td>
      <td>7.082959</td>
      <td>7.075700</td>
      <td>2.681833</td>
      <td>...</td>
      <td>1.137931</td>
      <td>0.001406</td>
      <td>0.510345</td>
      <td>0.001406</td>
      <td>21.809771</td>
      <td>0.057564</td>
      <td>True</td>
      <td>Adipocytes vs Rest</td>
      <td>Adipocytes</td>
      <td>Rest</td>
    </tr>
    <tr>
      <th>ADIPOQ</th>
      <td>0.9988</td>
      <td>0.0012</td>
      <td>6.724225</td>
      <td>0.003627</td>
      <td>0.000052</td>
      <td>0.0</td>
      <td>0.25</td>
      <td>7.722131</td>
      <td>7.461277</td>
      <td>3.332577</td>
      <td>...</td>
      <td>2.324136</td>
      <td>0.003622</td>
      <td>0.593103</td>
      <td>0.003352</td>
      <td>33.361748</td>
      <td>0.217763</td>
      <td>True</td>
      <td>Adipocytes vs Rest</td>
      <td>Adipocytes</td>
      <td>Rest</td>
    </tr>
    <tr>
      <th>GPAM</th>
      <td>0.9986</td>
      <td>0.0014</td>
      <td>6.569875</td>
      <td>0.025417</td>
      <td>0.000202</td>
      <td>0.0</td>
      <td>0.25</td>
      <td>7.365266</td>
      <td>7.381156</td>
      <td>2.562121</td>
      <td>...</td>
      <td>17.372416</td>
      <td>0.035791</td>
      <td>0.896552</td>
      <td>0.031520</td>
      <td>280.340485</td>
      <td>1.565905</td>
      <td>True</td>
      <td>Adipocytes vs Rest</td>
      <td>Adipocytes</td>
      <td>Rest</td>
    </tr>
    <tr>
      <th>PLIN1</th>
      <td>0.9984</td>
      <td>0.0016</td>
      <td>6.436144</td>
      <td>0.004482</td>
      <td>0.000048</td>
      <td>0.0</td>
      <td>0.25</td>
      <td>7.818194</td>
      <td>7.579515</td>
      <td>2.977385</td>
      <td>...</td>
      <td>2.799999</td>
      <td>0.004379</td>
      <td>0.806897</td>
      <td>0.004325</td>
      <td>52.921761</td>
      <td>0.196486</td>
      <td>True</td>
      <td>Adipocytes vs Rest</td>
      <td>Adipocytes</td>
      <td>Rest</td>
    </tr>
    <tr>
      <th>GPD1</th>
      <td>0.9974</td>
      <td>0.0026</td>
      <td>5.949637</td>
      <td>0.002172</td>
      <td>0.000044</td>
      <td>0.0</td>
      <td>0.25</td>
      <td>6.543847</td>
      <td>6.023436</td>
      <td>2.865962</td>
      <td>...</td>
      <td>1.758620</td>
      <td>0.007353</td>
      <td>0.620690</td>
      <td>0.007299</td>
      <td>30.203791</td>
      <td>0.255079</td>
      <td>True</td>
      <td>Adipocytes vs Rest</td>
      <td>Adipocytes</td>
      <td>Rest</td>
    </tr>
  </tbody>
</table>
<p>5 rows × 22 columns</p>
</div></div></div>
</div>
<p>We now extract top markers for each cluster using the DE results.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">markers</span> <span class="o">=</span> <span class="p">{}</span>
<span class="n">cats</span> <span class="o">=</span> <span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">cell_type</span><span class="o">.</span><span class="n">cat</span><span class="o">.</span><span class="n">categories</span>
<span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">c</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">cats</span><span class="p">):</span>
    <span class="n">cid</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="si">{}</span><span class="s2"> vs Rest&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">c</span><span class="p">)</span>
    <span class="n">cell_type_df</span> <span class="o">=</span> <span class="n">de_df</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">de_df</span><span class="o">.</span><span class="n">comparison</span> <span class="o">==</span> <span class="n">cid</span><span class="p">]</span>

    <span class="n">cell_type_df</span> <span class="o">=</span> <span class="n">cell_type_df</span><span class="p">[</span><span class="n">cell_type_df</span><span class="o">.</span><span class="n">lfc_mean</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">]</span>

    <span class="n">cell_type_df</span> <span class="o">=</span> <span class="n">cell_type_df</span><span class="p">[</span><span class="n">cell_type_df</span><span class="p">[</span><span class="s2">&quot;bayes_factor&quot;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mi">3</span><span class="p">]</span>
    <span class="n">cell_type_df</span> <span class="o">=</span> <span class="n">cell_type_df</span><span class="p">[</span><span class="n">cell_type_df</span><span class="p">[</span><span class="s2">&quot;non_zeros_proportion1&quot;</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mf">0.1</span><span class="p">]</span>

    <span class="n">markers</span><span class="p">[</span><span class="n">c</span><span class="p">]</span> <span class="o">=</span> <span class="n">cell_type_df</span><span class="o">.</span><span class="n">index</span><span class="o">.</span><span class="n">tolist</span><span class="p">()[:</span><span class="mi">3</span><span class="p">]</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">dendrogram</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;cell_type&quot;</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="s2">&quot;X_scVI&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">dotplot</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span> 
    <span class="n">markers</span><span class="p">,</span> 
    <span class="n">groupby</span><span class="o">=</span><span class="s1">&#39;cell_type&#39;</span><span class="p">,</span> 
    <span class="n">dendrogram</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> 
    <span class="n">color_map</span><span class="o">=</span><span class="s2">&quot;Blues&quot;</span><span class="p">,</span>
    <span class="n">swap_axes</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">use_raw</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">standard_scale</span><span class="o">=</span><span class="s2">&quot;var&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/api_overview_60_0.png" src="../_images/api_overview_60_0.png" />
</div>
</div>
<p>We can also visualize the scVI normalized gene expression values with the <code class="docutils literal notranslate"><span class="pre">layer</span></code> option.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">heatmap</span><span class="p">(</span>
    <span class="n">adata</span><span class="p">,</span> 
    <span class="n">markers</span><span class="p">,</span> 
    <span class="n">groupby</span><span class="o">=</span><span class="s1">&#39;cell_type&#39;</span><span class="p">,</span> 
    <span class="n">layer</span><span class="o">=</span><span class="s2">&quot;scvi_normalized&quot;</span><span class="p">,</span>
    <span class="n">standard_scale</span><span class="o">=</span><span class="s2">&quot;var&quot;</span><span class="p">,</span>
    <span class="n">dendrogram</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">8</span><span class="p">,</span> <span class="mi">12</span><span class="p">)</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<img alt="../_images/api_overview_62_0.png" src="../_images/api_overview_62_0.png" />
</div>
</div>
</div>
<div class="section" id="logging-information">
<h2>Logging information<a class="headerlink" href="#logging-information" title="Permalink to this headline">¶</a></h2>
<p>Verbosity varies in the following way:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">logger.setLevel(logging.WARNING)</span></code> will show a progress bar.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">logger.setLevel(logging.INFO)</span></code> will show global logs including the number of jobs done.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">logger.setLevel(logging.DEBUG)</span></code> will show detailed logs for each training (e.g the parameters tested).</p></li>
</ul>
<p>This function’s behaviour can be customized, please refer to its documentation for information about the different parameters available.</p>
<p>In general, you can use <code class="docutils literal notranslate"><span class="pre">scvi.settings.verbosity</span></code> to set the verbosity of the scvi package.
Note that <code class="docutils literal notranslate"><span class="pre">verbosity</span></code> corresponds to the logging levels of the standard python <code class="docutils literal notranslate"><span class="pre">logging</span></code> module. By default, that verbosity level is set to <code class="docutils literal notranslate"><span class="pre">INFO</span></code> (=20).
As a reminder the logging levels are:</p>
<table class="docutils align-center">
<colgroup>
<col style="width: 48%">
<col style="width: 52%">
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Level</p></th>
<th class="head"><p>Numeric value</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">CRITICAL</span></code></p></td>
<td><p>50</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">ERROR</span></code></p></td>
<td><p>40</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">WARNING</span></code></p></td>
<td><p>30</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">INFO</span></code></p></td>
<td><p>20</p></td>
</tr>
<tr class="row-even"><td><p><code class="docutils literal notranslate"><span class="pre">DEBUG</span></code></p></td>
<td><p>10</p></td>
</tr>
<tr class="row-odd"><td><p><code class="docutils literal notranslate"><span class="pre">NOTSET</span></code></p></td>
<td><p>0</p></td>
</tr>
</tbody>
</table></div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./introduction"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By Theislab<br/>
        
            &copy; Copyright 2021.<br/>
          <div class="extra_footer">
            <div>
  Brought to you by
  <a href="https://www.github.com/theislab">Theislab</a>, with many thanks to the single-cell community as a whole!
</div>

          </div>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>