
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>1. Prior art &#8212; Extended single-cell analysis tutorial</title>
    
  <link href="../_static/css/theme.css" rel="stylesheet">
  <link href="../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../_static/book.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
    <script src="../_static/jquery.js"></script>
    <script src="../_static/underscore.js"></script>
    <script src="../_static/doctools.js"></script>
    <script src="../_static/togglebutton.js"></script>
    <script src="../_static/clipboard.min.js"></script>
    <script src="../_static/copybutton.js"></script>
    <script src="../_static/book.js"></script>
    <script async="async" kind="hypothesis" src="https://hypothes.is/embed.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../_static/sphinx-thebe.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <link rel="shortcut icon" href="../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../genindex.html" />
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="2. A gentle introduction to single-cell RNA sequencing" href="experimental_data_collection.html" />
    <link rel="prev" title="Extended Single-cell Best-practices" href="../preamble.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../_static/logo.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">Extended single-cell analysis tutorial</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Introduction
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   1. Prior art
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="experimental_data_collection.html">
   2. A gentle introduction to single-cell RNA sequencing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="raw_data_processing.html">
   3. Raw data processing
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="analysis_tools.html">
   4. Computational single-cell analysis frameworks and tools
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="data_infrastructure.html">
   5. Data infrastructure
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="interoperability.html">
   6. Interoperability
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Preprocessing and visualization
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../preprocessing_visualization/introduction.html">
   7. Preprocessing &amp; visualization introduction
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Dealing with batch effects
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../batch_effects/introduction.html">
   8. Dealing with batch effects introduction
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Interpreting cellular structure
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../cellular_structure/introduction.html">
   9. Interpreting cellular structure introduction
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Dealing with conditions
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../conditions/introduction.html">
   10. Dealing with conditions introduction
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../conditions/differential_gene_expression.html">
   11. Differential gene expression analysis of human COVID-19 PBMCs
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Modeling mechanisms
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../mechanisms/introduction.html">
   12. Modeling mechanisms introduction
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Deconvolution
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../deconvolution/introduction.html">
   13. Deconvolution introduction
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Spatial omics
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../spatial/introduction.html">
   14. Spatial omics introduction
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multi-omics
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../multi_omics/introduction.html">
   15. Multi-omics introduction
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reproducibility
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../reproducibility/introduction.html">
   16. Reproducibility introduction
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Acknowledgements
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../acknowledgements.html">
   17. Acknowledgements
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Glossary
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../glossary.html">
   18. Glossary
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../_sources/introduction/prior_art.md"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.md</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/theislab/extended-single-cell-tutorial"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/theislab/extended-single-cell-tutorial/issues/new?title=Issue%20on%20page%20%2Fintroduction/prior_art.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/theislab/extended-single-cell-tutorial/edit/master/introduction/prior_art.md"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#bioconductor-osca">
   1.1. Bioconductor OSCA
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#current-best-practices-in-single-cell-rna-seq-analysis-a-tutorial">
   1.2. Current best practices in single-cell RNA-seq analysis: a tutorial
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#references">
   1.3. References
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="tex2jax_ignore mathjax_ignore section" id="prior-art">
<h1><span class="section-number">1. </span>Prior art<a class="headerlink" href="#prior-art" title="Permalink to this headline">¶</a></h1>
<p>Single-cell analysis is slowly transitioning from an interesting niche topic to a more mature field. Hence, we are arguably not the first to write a book on single-cell analysis, let alone guides and tutorials. In the following sections we highlight two existing and ongoing efforts to teach single-cell analysis and emphasize commonalities and differences to this book.</p>
<div class="section" id="bioconductor-osca">
<h2><span class="section-number">1.1. </span>Bioconductor OSCA<a class="headerlink" href="#bioconductor-osca" title="Permalink to this headline">¶</a></h2>
<p>Orchestrating Single-Cell Analysis with Bioconductor (Bioconductor OSCA)<span id="id1">[<a class="reference internal" href="#id32">Amezquita <em>et al.</em>, 2022</a>]</span> accessible at <a class="reference external" href="https://bioconductor.org/books/release/OSCA/">https://bioconductor.org/books/release/OSCA/</a> is a digital book which aims to teach common workflows for the analysis of single-cell RNA-Seq with the R based Bioconductor<span id="id2">[<a class="reference internal" href="#id7">Huber <em>et al.</em>, 2015</a>]</span> ecosystem. A paper with the same name<span id="id3">[<a class="reference internal" href="#id33">Amezquita <em>et al.</em>, 2020</a>]</span> presented an overview of single-cell analysis with Bioconductor and the book is an associated online version which goes into greater detail with extensive code examples.</p>
<p>The book is very comprehensive with respect to basic single-cell RNA-Seq analysis with great explanations and extensive workflow examples. However, it does not comprise advanced topics such as, for example, RNA velocity or spatial transcriptomics. Moreover, additional modalities such as ATAC-Seq or CITE-Seq data, or the multimodal integration of these are not covered either. Since the book is designed for the Bioconductor ecosystem it only employs the in Bioconductor available tools. As denoted in the book itself, these do not necessarily result in an optimal analysis. We perceive the Bioconductor OSCA book as especially useful for people with a basic R and stronger biology background who are interested in learning how to analyze single-cell RNA-Seq data analysis with Bioconductor.</p>
</div>
<div class="section" id="current-best-practices-in-single-cell-rna-seq-analysis-a-tutorial">
<h2><span class="section-number">1.2. </span>Current best practices in single-cell RNA-seq analysis: a tutorial<a class="headerlink" href="#current-best-practices-in-single-cell-rna-seq-analysis-a-tutorial" title="Permalink to this headline">¶</a></h2>
<p>Current best practices in single-cell RNA-seq analysis: a tutorial<span id="id4">[<a class="reference internal" href="../conditions/differential_gene_expression.html#id51">Luecken and Theis, 2019</a>]</span> is a paper written by Malte Lücken and Fabian Theis which introduces best practice single-cell RNA-Seq analysis. The unique contribution of the paper to the field is that it not only serves as a review of the possible analysis steps, but always suggests best-practices based on unbiased benchmarks. Whenever recommendations for best-practices are not available, general recommendations for analysis approaches are suggested. The paper itself is accompanied with an <a class="reference external" href="https://github.com/theislab/single-cell-tutorial/">example analysis of mouse intestinal epithelium regions</a> from Haber et al<span id="id5">[<a class="reference internal" href="#id35">Haber <em>et al.</em>, 2017</a>]</span>.</p>
<p>Compared to Bioconductor OSCA, the paper and the example analysis is not biased by the tools that it showcases and more complete in content with respect to the breadth of covered topics. Nevertheless, the associated example analysis lacks in newbie friendliness and has already become outdated. Moreover, similarly to the Bioconductor OSCA paper and book, Lücken and Theis do not cover more recent topics such as RNA velocity, spatial transcriptomics or multi-omics. We strongly recommend the paper as an introduction and overview to the field and initial analysis best-practice recommendations. The chapters in this book are based on the most recent best-practices and provide an updated view on the field. We generally advise against examining the associated case-study and suggest to instead read the chapters of this book in detail.</p>
</div>
<div class="section" id="references">
<h2><span class="section-number">1.3. </span>References<a class="headerlink" href="#references" title="Permalink to this headline">¶</a></h2>
<p id="id6"><dl class="citation">
<dt class="label" id="id32"><span class="brackets"><a class="fn-backref" href="#id1">ALHR22</a></span></dt>
<dd><p>Robert Amezquita, Aaron Lun, Stephanie Hicks, and Gottardo Raphael. Orchestrating single-cell analysis with bioconductor. <span><a class="reference external" href="#"></a></span>https://bioconductor.org/books/release/OSCA/, 2022. Accessed: 2022-04-21.</p>
</dd>
<dt class="label" id="id33"><span class="brackets"><a class="fn-backref" href="#id3">ALB+20</a></span></dt>
<dd><p>Robert A. Amezquita, Aaron T. L. Lun, Etienne Becht, Vince J. Carey, Lindsay N. Carpp, Ludwig Geistlinger, Federico Marini, Kevin Rue-Albrecht, Davide Risso, Charlotte Soneson, Levi Waldron, Hervé Pagès, Mike L. Smith, Wolfgang Huber, Martin Morgan, Raphael Gottardo, and Stephanie C. Hicks. Orchestrating single-cell analysis with bioconductor. <em>Nature Methods</em>, 17(2):137–145, Feb 2020. URL: <a class="reference external" href="https://doi.org/10.1038/s41592-019-0654-x">https://doi.org/10.1038/s41592-019-0654-x</a>, <a class="reference external" href="https://doi.org/10.1038/s41592-019-0654-x">doi:10.1038/s41592-019-0654-x</a>.</p>
</dd>
<dt class="label" id="id35"><span class="brackets"><a class="fn-backref" href="#id5">HBR+17</a></span></dt>
<dd><p>Adam L. Haber, Moshe Biton, Noga Rogel, Rebecca H. Herbst, Karthik Shekhar, Christopher Smillie, Grace Burgin, Toni M. Delorey, Michael R. Howitt, Yarden Katz, Itay Tirosh, Semir Beyaz, Danielle Dionne, Mei Zhang, Raktima Raychowdhury, Wendy S. Garrett, Orit Rozenblatt-Rosen, Hai Ning Shi, Omer Yilmaz, Ramnik J. Xavier, and Aviv Regev. A single-cell survey of the small intestinal epithelium. <em>Nature</em>, 551(7680):333–339, Nov 2017. URL: <a class="reference external" href="https://doi.org/10.1038/nature24489">https://doi.org/10.1038/nature24489</a>, <a class="reference external" href="https://doi.org/10.1038/nature24489">doi:10.1038/nature24489</a>.</p>
</dd>
<dt class="label" id="id7"><span class="brackets"><a class="fn-backref" href="#id2">HCG+15</a></span></dt>
<dd><p>Wolfgang Huber, Vincent J. Carey, Robert Gentleman, Simon Anders, Marc Carlson, Benilton S. Carvalho, Hector Corrada Bravo, Sean Davis, Laurent Gatto, Thomas Girke, Raphael Gottardo, Florian Hahne, Kasper D. Hansen, Rafael A. Irizarry, Michael Lawrence, Michael I. Love, James MacDonald, Valerie Obenchain, Andrzej K. Oleś, Hervé Pagès, Alejandro Reyes, Paul Shannon, Gordon K. Smyth, Dan Tenenbaum, Levi Waldron, and Martin Morgan. Orchestrating high-throughput genomic analysis with bioconductor. <em>Nature Methods</em>, 12(2):115–121, Feb 2015. URL: <a class="reference external" href="https://doi.org/10.1038/nmeth.3252">https://doi.org/10.1038/nmeth.3252</a>, <a class="reference external" href="https://doi.org/10.1038/nmeth.3252">doi:10.1038/nmeth.3252</a>.</p>
</dd>
<dt class="label" id="id34"><span class="brackets"><a class="fn-backref" href="#id4">LT19</a></span></dt>
<dd><p>Malte D Luecken and Fabian J Theis. Current best practices in single-cell rna-seq analysis: a tutorial. <em>Molecular Systems Biology</em>, 15(6):e8746, 2019. URL: <a class="reference external" href="https://www.embopress.org/doi/abs/10.15252/msb.20188746">https://www.embopress.org/doi/abs/10.15252/msb.20188746</a>, <a class="reference external" href="https://arxiv.org/abs/https://www.embopress.org/doi/pdf/10.15252/msb.20188746">arXiv:https://www.embopress.org/doi/pdf/10.15252/msb.20188746</a>, <a class="reference external" href="https://doi.org/https://doi.org/10.15252/msb.20188746">doi:https://doi.org/10.15252/msb.20188746</a>.</p>
</dd>
</dl>
</p>
</div>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./introduction"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="../preamble.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Extended Single-cell Best-practices</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="experimental_data_collection.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title"><span class="section-number">2. </span>A gentle introduction to single-cell RNA sequencing</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By Theislab<br/>
        
            &copy; Copyright 2021.<br/>
          <div class="extra_footer">
            <div>
  Brought to you by
  <a href="https://www.github.com/theislab">Theislab</a>, with many thanks to the single-cell community as a whole!
</div>

          </div>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>